cloudstack/resource_cloudstack_service_offering.go

```diff
@@ -20,246 +20,735 @@
+	}
 
-	if err != nil {
-		return err
+	if !data.DeploymentPlanner.IsNull() && !data.DeploymentPlanner.IsUnknown() {
+		p.SetDeploymentplanner(data.DeploymentPlanner.ValueString())
+	} else {
+		data.DeploymentPlanner = types.StringNull()
 	}
 
-	log.Printf("[DEBUG] Service Offering %s successfully created", name)
-	d.SetId(s.Id)
+	if !data.ZoneId.IsNull() && !data.ZoneId.IsUnknown() {
+		zoneIds := make([]string, 0, len(data.ZoneId.Elements()))
+		diags := data.ZoneId.ElementsAs(ctx, &zoneIds, false)
+		if diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+		p.SetZoneid(zoneIds)
+	} else {
+		data.ZoneId = types.ListNull(types.StringType)
+	}
 
-	return resourceCloudStackServiceOfferingRead(d, meta)
-}
+	if !data.DiskOfferingId.IsNull() && !data.DiskOfferingId.IsUnknown() {
+		p.SetDiskofferingid(data.DiskOfferingId.ValueString())
+	} else {
+		p.SetDiskofferingid("")
+		data.DiskOfferingId = types.StringNull()
+
+		if !data.StorageType.IsNull() && !data.StorageType.IsUnknown() {
+			p.SetStoragetype(data.StorageType.ValueString())
+		} else {
+			data.StorageType = types.StringNull()
+		}
 
-func resourceCloudStackServiceOfferingRead(d *schema.ResourceData, meta interface{}) error {
-	cs := meta.(*cloudstack.CloudStackClient)
-	log.Printf("[DEBUG] Retrieving Service Offering %s", d.Get("name").(string))
+		if !data.ProvisioningType.IsNull() && !data.ProvisioningType.IsUnknown() {
+			p.SetProvisioningtype(data.ProvisioningType.ValueString())
+		}
 
-	// Get the Service Offering details
-	s, count, err := cs.ServiceOffering.GetServiceOfferingByName(d.Get("name").(string))
+		if !data.WriteCacheType.IsNull() && !data.WriteCacheType.IsUnknown() {
+			p.SetCachemode(data.WriteCacheType.ValueString())
+		}
 
-	if err != nil {
-		if count == 0 {
-			log.Printf("[DEBUG] Service Offering %s does no longer exist", d.Get("name").(string))
-			d.SetId("")
-			return nil
+		if data.QosType.ValueString() == "hypervisor" {
+			if !data.DiskReadRateBps.IsNull() && !data.DiskReadRateBps.IsUnknown() {
+				p.SetBytesreadrate(data.DiskReadRateBps.ValueInt64())
+			}
+
+			if !data.DiskWriteRateBps.IsNull() && !data.DiskWriteRateBps.IsUnknown() {
+				p.SetByteswriterate(data.DiskWriteRateBps.ValueInt64())
+			}
+
+			if !data.DiskReadRateIops.IsNull() && !data.DiskReadRateIops.IsUnknown() {
+				p.SetIopsreadrate(data.DiskReadRateIops.ValueInt64())
+			}
+
+			if !data.DiskWriteRateIops.IsNull() && !data.DiskWriteRateIops.IsUnknown() {
+				p.SetIopswriterate(data.DiskWriteRateIops.ValueInt64())
+			}
+		} else if data.QosType.ValueString() == "storage" {
+			p.SetCustomizediops(data.CustomIops.ValueBool())
+
+			if !data.CustomIops.ValueBool() {
+				if !data.MinIops.IsNull() && !data.MinIops.IsUnknown() {
+					p.SetMiniops(data.MinIops.ValueInt64())
+				}
+
+				if !data.MaxIops.IsNull() && !data.MaxIops.IsUnknown() {
+					p.SetMaxiops(data.MaxIops.ValueInt64())
+				}
+			} else {
+				p.SetMiniops(0)
```

Comment:
I am not sure. But this could cause issues. I would suggest not doing anything if it's not defined.